### PR TITLE
Add validation to image sharing

### LINKED ISSUE #4: Image URL Validation
#### As a user I want to be prevented from submitting invalid image URLs.

__Story__:

We quickly got up to speed providing our users the ability to add images to our
service. How cool are all the images they are adding? But wait, what are all
these broken image placeholders? We would like to avoid that. In order to do so
we will need to add some validation to our model in order to increase the
chance that the data submitted actually corresponds to an image URL.

__Acceptance criteria__:
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [x] What is a valid image URL?

__Dependencies__:
- Save Image Link

### PR DESCRIPTION
**Pull request by mathewbaltes**
Closes #4

### CODE REVIEW

File: app/models/image.rb [hunk 1]
```diff
@@ -1,3 +1,12 @@
+require 'uri'
+
+class UrlValidator < ActiveModel::Validator
```
Comment (ptevis42) [inline app/models/image.rb]: Preferred style is for Validators to live in their own file, separate from the Model (following the "one class per file" approach that Rails favors). That file should probably live in `app/validators`.

File: test/models/images_model_test.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+require 'test_helper'
+
+class Model < ActiveSupport::TestCase
+  test 'validate model behaviors' do
+    assert_equal true, Image.create(url: 'http://www.appfolio.com/').valid?
```
Comment (ptevis42) [inline test/models/images_model_test.rb]: You should also test the `UrlValidator` directly, asserting that it sets error messages correctly (and the content of those messages). That test should probably live in `tests/validators`.

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -13,6 +13,9 @@ def create
     @image.save!
     flash[:success] = 'Thank you for sharing your image with us! The image you have shared is below.'
     redirect_to @image
+  rescue ActiveRecord::RecordInvalid
```
Comment (ptevis42) [inline app/controllers/images_controller.rb]: Is failing a validation an expected case, or a truly exceptional one? If the former, we shouldn't use exception handling for that kind of control flow.

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -13,6 +13,9 @@ def create
...
+    flash[:error] = 'Please insert a valid url.'
```
Comment (ptevis42) [inline app/controllers/images_controller.rb]: Why is the content of the error message specified in the controller? It shouldn't know what kind of error has occurred; the validator does.

File: test/validators/url_validator_test.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+require 'test_helper'
+
+class Model < ActiveSupport::TestCase
+  test 'validate proper url validation' do
+    assert_equal ['Need to input valid url.'], UrlValidator.new.validate(Image.new(url: ''))
```
Comment (ptevis42) [inline test/validators/url_validator_test.rb]: Is there a way to break the explicit dependency of this test on Image? Can you stub/mock/duck-type something so that you don't refer explicitly to its type?

File: test/validators/url_validator_test.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+require 'test_helper'
+
+class Model < ActiveSupport::TestCase
```
Comment (ptevis42) [inline test/validators/url_validator_test.rb]: The usual pattern is to name Tests after the class they are testing (i.e. `UrlValidatorTest`). That also keeps the name consistent with the filename.

File: test/validators/url_validator_test.rb (hunk 5 above)
Comment (ptevis42) [inline test/validators/url_validator_test.rb]: The expected behavior for a validator is that it's going to update the object that it's running the validation on. You're asserting that the last line of the `validate` method is assigning the error string to something (and implicitly returning that value). What you really should be asserting is something about  `errors.messages[:url]`.

File: test/validators/url_validator_test.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+  test 'validate proper url validation' do
+    assert_equal ['Need to input valid url.'], UrlValidator.new.validate(Image.new(url: ''))
+    assert_equal ['Need to input valid url.'], UrlValidator.new.validate(Image.new(url: 'blah'))
```
Comment (ptevis42) [inline test/validators/url_validator_test.rb]: I suspect you'll want to break this into multiple tests (possibly with a `setup`).

File: app/models/image.rb [hunk 8]
```diff
@@ -1,12 +1,4 @@
...
-class UrlValidator < ActiveModel::Validator
-  def validate(record)
-    record.errors[:url] << 'Need to input valid url.' unless record.url =~ URI.regexp
-  end
-end
-
 class Image < ApplicationRecord
   include ActiveModel::Validations
-  validates :url, presence: true, url: true
+  validates :url, url: true
```
Comment (ptevis42) [inline app/models/image.rb]: Why did you remove the `presence` validation? Does something else cover that?

File: app/controllers/images_controller.rb [hunk 9]
```diff
@@ -10,12 +10,14 @@ def new
...
-    redirect_to @image
-  rescue ActiveRecord::RecordInvalid
-    flash[:error] = 'Please insert a valid url.'
-    redirect_to new_image_path
+    if @image.valid?
+      @image.save!
+      flash[:success] = 'Thank you for sharing your image with us! The image you have shared is below.'
+      redirect_to @image
+    else
+      flash[:error] = @image.errors.messages[:url].last
```
Comment (ptevis42) [inline app/controllers/images_controller.rb]: This line means you will only ever report url validation errors, and you will only report the last one. Is there a way to keep the `create` action from having to know that errors the model could generate?

File: app/models/image.rb (hunk 8 above)
Comment (mathewbaltes) [inline app/models/image.rb]: The url validator handles validation for urls.

File: test/validators/url_validator_test.rb (hunk 7 above)
Comment (mathewbaltes) [inline test/validators/url_validator_test.rb]: Bad habits from the old days. I will break them up.

File: app/views/images/new.html.erb [hunk 10]
```diff
@@ -1,14 +1,19 @@
 <%= form_for :image, url: images_path do |f| %>
   <div class="row justify-content-center">
     <div class="col-lg-6">
-      <div class="form-group">
+      <div class="form-group <%= 'has-danger' if flash[:error] %>">
           <%= f.text_field :url, class: 'form-control', id: 'link-input' %>
+          <% if flash[:error] %>
+            <% flash[:error]['url'].each do |url_error| %>
```
Comment (ptevis42) [inline app/views/images/new.html.erb]: Can you do this for all errors, rather than just `url` errors? That way the view doesn't have to know anything about validation.

File: app/models/image.rb [hunk 11]
```diff
@@ -1,3 +1,4 @@
 class Image < ApplicationRecord
-  validates :url, presence: true
+  include ActiveModel::Validations
+  validates :url, url: true
```
Comment (ptevis42) [inline app/models/image.rb]: Is there something that handles the presence validation?

File: test/models/images_model_test.rb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+require 'test_helper'
+
+class ImageModelTest < ActiveSupport::TestCase
+  test 'verify image with valid url is valid' do
+    assert_equal true, Image.create(url: 'http://www.appfolio.com/').valid?
```
Comment (ptevis42) [inline test/models/images_model_test.rb]: I would probably verify that that `errors` is empty.

File: test/models/images_model_test.rb [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  end
+  test 'verify image with invalid url is invalid' do
+    assert_equal false, Image.create(url: 'blah').valid?
```
Comment (ptevis42) [inline test/models/images_model_test.rb]: In this and the following case, I would probably verify that `errors` is non-empty (but not the specifics of what is in there).

File: app/models/image.rb (hunk 11 above)
Comment (mathewbaltes) [inline app/models/image.rb]: Yes, the URL validator handles presence validation.
